### PR TITLE
WIP Migrate to styled-system v5

### LINKED ISSUE #472: Upgrade to styled-system v5
@jxnblk just released some cool new APIs and performance improvements in styled-system v5. We should upgrade :)

[Migration guide](https://styled-system.com/guides/migrating)

### PR DESCRIPTION
**Pull request by jxnblk**
Attempts to update to `styled-system` version 5.

- Prop types were removed from the core library and are now imported from `@styled-system/prop-types`
- Currently, this consolidates all flexbox style props together, instead of using separate flex container and item props. This might not be something you want, so I've also included a code comment with an example of keeping the current props API in tact
- Replaces `width` with the `layout` category in Button (again, this may not be what you want and can be changed)
- Removes `system-components` from dependencies – it didn't look like this was in use

**To do:**

- [ ] Tests for the `Avatar` component are failing, and I haven't spent any time looking into that yet
- [ ] The border color in the docs sidebar looks different (this might be a bug with `styled-system`)

Closes #472

- [ ] Updated README

#### Merge checklist
- [ ] Changed base branch to release branch
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

### CODE REVIEW

**Comment by emplums** [comment]
THANK YOU!!! 🙏 🙏 🙏 I will take a look at this today!

**Comment by jxnblk** [comment]
😄 No rush! Feel free to change whatever – this felt like a great place to test out migration and make sure things still worked

File: src/constants.js [hunk 1]
```diff
@@ -1,75 +1,65 @@
...
+import systemPropTypes from '@styled-system/prop-types'
+import themeGet from '@styled-system/theme-get'
 import theme from './theme'
 
-const {compose, get: getKey, themeGet} = styledSystem
+const { get: getKey } = styledSystem
 
 export const get = key => themeGet(key, getKey(theme, key))
 
-export const COMMON_LIST = ['color', 'space']
```
Comment (colebemis) [inline src/constants.js]: We used these lists to document system props here: https://github.com/primer/components/blob/master/pages/components/docs/system-props.md

How would you recommend generating these lists for documentation? With the addition of style categories, maybe our system prop groups are simple enough that we can keep them up-to-date manually? @emplums We could potentially link directly to styled-system's reference table https://styled-system.com/table

File: src/constants.js [hunk 2]
```diff
@@ -1,75 +1,65 @@
 import * as styledSystem from 'styled-system'
+import { compose } from 'styled-system'
+import systemPropTypes from '@styled-system/prop-types'
+import themeGet from '@styled-system/theme-get'
 import theme from './theme'
 
-const {compose, get: getKey, themeGet} = styledSystem
+const { get: getKey } = styledSystem
```
Comment (colebemis) [inline src/constants.js]: We could probably combine this with the `compose` import a few lines above:

```js
import * as styledSystem from 'styled-system'
import {compose, get} from 'styled-system'
```

File: src/constants.js [hunk 3]
```diff
@@ -1,75 +1,65 @@
...
-  'height',
-  'minWidth',
-  'minHeight',
-  'maxWidth',
-  'maxHeight',
-  'overflow',
-  'verticalAlign'
-]
+// these are 1:1 with styled-system's API now,
+// so you could consider dropping the abstraction
```
Comment (colebemis) [inline src/constants.js]: 🎉 I'm super open to importing `typography`, `layout`, etc. directly from `styled-system`

File: src/constants.js [hunk 4]
```diff
@@ -1,75 +1,65 @@
...
-  'fontFamily',
-  'fontSize',
-  'fontStyle',
-  'fontWeight',
-  'lineHeight',
-  'textAlign'
-]
+export const COMMON = compose(
+  styledSystem.space,
+  styledSystem.color,
```
Comment (colebemis) [inline src/constants.js]: I think we recently added `display` to this list.

File: src/constants.js (hunk 1 above)
Comment (jxnblk) [reply src/constants.js]: Currently undocumented but each “category” from Styled System has a `.propNames` array, which hopefully you could leverage for this

Comment (colebemis) [reply src/constants.js]: Oh nice. Will that work with `compose` too?

Comment (jxnblk) [reply src/constants.js]: Yup. If you're interested in the implementation, the relevant parts are here: https://github.com/styled-system/styled-system/blob/master/packages/core/src/index.js#L68

Each call to `system` uses the `createParser` to create a "parser" function which includes the arguments as a `.config` property – and also adds a `.propNames` property. The `compose` utility uses each function's `.config` property to create a new parser function based on that config. This is intended to be a mechanism not only for internal use, but also as a better way to generate docs for components built with Styled System, and I'd like to add some more docs and utilities for this stuff in the future

**Comment by emplums** [comment]
Finally getting around to looking at this after being out a while for vacation. Going to make a to do list here for things I'd like to update before merging which I'll plan on doing myself this week.

### To Do
- [ ] Reimplement prop category documentation
- [ ] Fix Sidebar border
- [ ] Fix Avatar tests

File: src/constants.js (hunk 4 above)
Comment (emplums) [reply src/constants.js]: yes we did but it hasn't merged yet. I'll make that change here too so it merges into the release branch fine

**Comment by emplums** [comment]
Going to merge this into the release branch, then make a separate PR for the changes listed above!

**Pull request merged**